sdk/python/kfp/dsl/artifact.py

```diff
@@ -98,129 +46,105 @@ class Artifact(object):
+  def set_custom_metadata_field(self, name: str, value: Any):
+    if name in self._metadata_fields:
+      raise RuntimeError("Schema field: {} cannot be set through set_custom_metadata_field.".format(name))
+    self.metadata[name] = value
+    metadata_protobuf_struct = Struct()
+    metadata_protobuf_struct.update(self.metadata)
+    self._artifact.metadata.CopyFrom(metadata_protobuf_struct)
+  
+  def get_custom_metadata_field(self, name: str) -> Any:
+    if name not in self.metadata:
+      raise RuntimeError("Metadata field: {} not set.", name)
+    return self.metadata[name]
 
   @property
   def type_schema(self) -> str:
-    """Gets the instance_schema specified for this Artifact object."""
-    return self._type_schema
-
-  def __repr__(self) -> str:
-    return 'Artifact(artifact: {}, type_schema: {})'.format(
-        str(self._artifact), str(self.type_schema))
+    """Gets the instance_schema for this Artifact object."""
+    return self._instance_schema
 
   def __getattr__(self, name: str) -> Any:
-    """Custom __getattr__ to allow access to artifact properties."""
-    if name == '_artifact_type':
-      # Prevent infinite recursion when used with copy.deepcopy().
-      raise AttributeError()
-    properties = self.PROPERTIES or {}
-    if name not in properties:
+    """Custom __getattr__ to allow access to artifact metadata."""
+
+    if name not in self._metadata_fields:
       raise AttributeError(
-          '%s artifact has no property %r.' % (self.TYPE_NAME, name))
-    property_type = properties[name].type
-    if property_type == PropertyType.STRING:
-      if name not in self._artifact.properties:
-        # Avoid populating empty property protobuf with the [] operator.
-        return ''
-      return self._artifact.properties[name].string_value
-    elif property_type == PropertyType.INT:
-      if name not in self._artifact.properties:
-        # Avoid populating empty property protobuf with the [] operator.
-        return 0
-      return self._artifact.properties[name].int_value
-    elif property_type == PropertyType.DOUBLE:
-      if name not in self._artifact.properties:
-        # Avoid populating empty property protobuf with the [] operator.
-        return 0.0
-      return self._artifact.properties[name].double_value
-    else:
-      raise Exception('Unknown MLMD type %r for property %r.' %
-                      (property_type, name))
+          'No metadata field: {} in artifact.'.format(name))
+
+    return self.metadata[name]
 
   def __setattr__(self, name: str, value: Any):
-    """Custom __setattr__ to allow access to artifact properties."""
+    """Custom __setattr__ to allow access to artifact metadata."""
+
     if not self._initialized:
       object.__setattr__(self, name, value)
       return
-    properties = self.PROPERTIES or {}
-    if name not in properties:
+    
+    metadata_fields = self._metadata_fields or {}
+    if name not in self._metadata_fields:
```

Comment:
if `self._metadata_fields` could be `None`, then we need the `or {}` in on both line 146 and 147. Otherwise, we don't need it on line 146.

Reply:
Fixed. Please let me know if this is what you meant.